src/boot/efi/boot.c

```diff
@@ -2268,6 +2306,104 @@ static VOID config_write_entries_to_variable(Config *config) {
         (void) efivar_set_raw(LOADER_GUID, L"LoaderEntries", buffer, (UINT8 *) p - (UINT8 *) buffer, 0);
 }
 
+static EFI_STATUS secure_boot_enroll_from(EFI_FILE_HANDLE dir) {
+        static const UINT32 sb_vars_opts =
+                EFI_VARIABLE_NON_VOLATILE |
+                EFI_VARIABLE_BOOTSERVICE_ACCESS |
+                EFI_VARIABLE_RUNTIME_ACCESS |
+                EFI_VARIABLE_TIME_BASED_AUTHENTICATED_WRITE_ACCESS |
+                0;
+
+        static const struct {
+                const CHAR16 *name;
+                const CHAR16 *filename;
+                const EFI_GUID vendor;
+        } sb_vars[] = {
+                { L"db", L"db.auth", EFI_IMAGE_SECURITY_DATABASE_VARIABLE },
+                { L"KEK", L"KEK.auth", EFI_GLOBAL_VARIABLE },
+                { L"PK", L"PK.auth", EFI_GLOBAL_VARIABLE },
+        };
+
+        _cleanup_freepool_ CHAR8 *buffer = NULL;
+        UINTN size;
+        EFI_STATUS err = EFI_SUCCESS;
+
+        for (UINTN i = 0; i < ELEMENTSOF(sb_vars); i++) {
+                file_read(dir, sb_vars[i].filename, 0, 0, &buffer, &size);
+
+                if (!EFI_ERROR(err)) {
+                        err = efivar_set_raw(&sb_vars[i].vendor, sb_vars[i].name, buffer, size, sb_vars_opts);
+                        if (EFI_ERROR(err)) {
+                                Print(L"Failed to write %s secure boot variable: %r\n", sb_vars[i].name, err);
+                                uefi_call_wrapper(BS->Stall, 1, 3 * 1000 * 1000);
+                                return err;
+                        }
+                } else {
+                        Print(L"Failed reading %s file: %r\n", sb_vars[i].filename, err);
+                        uefi_call_wrapper(BS->Stall, 1, 3 * 1000 * 1000);
+                        return err;
+                }
+        }
+
+        Print(L"Finished loading secure boot variables.\n");
+        uefi_call_wrapper(BS->Stall, 1, 3 * 1000 * 1000);
+        return err;
+}
```

Comment:
the way the function is implemented now it it will fail on every error condition and interleave file read and efi var write.

I mean, I think two policies would make sense: log error conditions but go on, executing as much as we can (which i proposed earlier), or be defensive and thus validate as much as we can beforehand and only then make changes. the latter would mean we'd read the three files first, and then push them into efi vars. 

I am fine with either logic, you pick. But i think the one implemented by the patch right now is not great, given that we'll only notice we have no PK after we already fucked up the db...